src/components/filters/useOperatingSystemFilter.js

```diff
@@ -1,37 +1,88 @@
 
-export const operatingSystemFilterState = { operatingSystemFilter: [] };
 export const OPERATING_SYSTEM_FILTER = 'OPERATING_SYSTEM_FILTER';
 export const operatingSystemFilterReducer = (_state, { type, payload }) => ({
     ...type === OPERATING_SYSTEM_FILTER && {
         operatingSystemFilter: payload
     }
 });
 
-export const useOperatingSystemFilter = (
-    operatingSystems,
-    areOperatingSystemsLoaded,
-    [state, dispatch] = [operatingSystemFilterState]
-) => {
-    // FIXME: is this state value needed and used at all?
-    const [operatingSystemStateValue, setStateValue] = useState([]);
-    const operatingSystemValue = dispatch ? state.operatingSystemFilter : operatingSystemStateValue;
-    const setValue = dispatch
-        ? (newValue) =>
-            dispatch({
-                type: OPERATING_SYSTEM_FILTER,
-                payload: Array.isArray(newValue)
-                    ? toGroupSelectionValue(newValue)
-                    : newValue
-            })
-        : setStateValue;
-    const filter = buildOSFilterConfig({
-        value: operatingSystemValue,
-        onChange: (_e, value) => setValue(value)
-        // TODO: better handling (spinner, skeleton) of loading systems?
-    }, areOperatingSystemsLoaded ? operatingSystems : []);
-    const chip = buildOSChip(operatingSystemValue,  areOperatingSystemsLoaded ? operatingSystems : []);
-
-    return [filter, chip, operatingSystemValue, setValue];
+/**
+* OS version filter hook.
+* @param {Array} apiParams - an array containing parameters for GET /system_profile/operating_system call
+* @return {Array} An array containing config object, chips array and value setter function.
+*/
+const useOperatingSystemFilter = (apiParams = []) => {
+    const dispatch = useDispatch();
+    const operatingSystems = useSelector(({ entities }) => entities?.operatingSystems);
+    const operatingSystemsLoaded = useSelector(({ entities }) => entities?.operatingSystemsLoaded) || false;
+    const availableVersions = operatingSystems === undefined ? [] : operatingSystems;
+
+    // selected versions has the boolean set to true
+    const [selected, setSelected] = useState({});
+    const [groups, setGroups] = useState([]);
+
+    useEffect(() => {
+        dispatch(fetchOperatingSystems(apiParams));
+    }, []);
+
+    useEffect(() => {
+        setGroups(groupOSFilterVersions(availableVersions));
+        setSelected((availableVersions).reduce((acc, { value: version }) => {
+            const [major] = version.split('.');
+            // use the previous value if avalable
+            set(acc, [major, version], get(selected, [major, version]) || false);
+            // update group selection identifier
+            set(acc, [major, major], Object.values({ ...acc[major] }).filter(v => v !== major).every(Boolean));
+            return acc;
+        }, {}));
+    }, [operatingSystems]);
+
+    // PrimaryToolbar filter configuration
+    const config = useMemo(() => ({
```

Comment:
needed to wrap couple of things in useMemo: because if declared with simple assignment `=`, this could lead to infinite reloading of tables that use this hook (e.g., some of the hook's return values can be added to table's useEffect dependencies)